engine/src/main/java/org/pentaho/di/trans/streaming/common/BlockingQueueStreamSource.java

```diff
@@ -0,0 +1,158 @@
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************************/
+
+
+package org.pentaho.di.trans.streaming.common;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.reactivex.Flowable;
+import io.reactivex.processors.PublishProcessor;
+import org.pentaho.di.core.logging.LogChannel;
+import org.pentaho.di.core.row.ValueMetaInterface;
+import org.pentaho.di.trans.streaming.api.StreamSource;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.pentaho.di.i18n.BaseMessages.getString;
+
+/**
+ * Implementation of StreamSource which handles pause/resume logic, as well as creation of .rows() which generates a
+ * blocking iterable.
+ * <p>
+ * Child classes should implement {@link #open()} to connect to a datasource.  The child .open implementation will
+ * typically start a new thread that feeds rows of data to the {@link #acceptRows(List)} method. Any resource cleanup
+ * should be included in a .close() implementation, along with a call to super.close() to complete the blocking
+ * iterable.
+ */
+public abstract class BlockingQueueStreamSource<T> implements StreamSource<T> {
+
+  private static final Class<?> PKG = BlockingQueueStreamSource.class;
+
+
+  private final AtomicBoolean paused = new AtomicBoolean( false );
+
+  private final PublishProcessor<T> publishProcessor = PublishProcessor.create();
+  protected final BaseStreamStep streamStep;
+
+  // binary semaphore used to block acceptance of rows when paused
+  @VisibleForTesting Semaphore acceptingRowsSemaphore = new Semaphore( 1 );
+  @VisibleForTesting LogChannel logChannel = new LogChannel( this );
+
+  protected BlockingQueueStreamSource( BaseStreamStep streamStep ) {
+    this.streamStep = streamStep;
+  }
+
+
+  @Override public Flowable<T> flowable() {
+    return Flowable
+      .fromPublisher( publishProcessor );
+  }
+
+  @Override public void close() {
+    if ( !publishProcessor.hasComplete() ) {
+
+      publishProcessor.onComplete();
+    }
+  }
+
+  /**
+   * Marks the source paused (if not already) and acquires the permit, which will cause acceptRows to block.
+   */
+  @Override public synchronized void pause() {
+    if ( !paused.getAndSet( true ) ) {
+      try {
+        assert acceptingRowsSemaphore.availablePermits() == 1;
+        acceptingRowsSemaphore.acquire();
+      } catch ( InterruptedException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)